Undertow Websockets: Could not create an endpoint dynamically

### Describe the bug

Extensions: `io.quarkus:quarkus-undertow-websockets`

On version 1.11.7.Final I could create an endpoint dynamically as follow:

```
 ((ServerContainer) sce.getServletContext().getAttribute(ServerContainer.class.getName()))
                    .addEndpoint(ServerEndpointConfig.Builder.create(WebsockEndpoint.class, "/simple")
                            .configurator(new WebsockEndpointConfigurator(this)).build());
```

However after upgrade to version `1.13.7.Final` I got a `NullPointerException`

```
Caused by: java.lang.RuntimeException: java.lang.RuntimeException: java.lang.NullPointerException
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder.bootServletContainer(UndertowDeploymentRecorder.java:528)
	at io.quarkus.deployment.steps.UndertowBuildStep$build-649634386.deploy_0(UndertowBuildStep$build-649634386.zig:238)
	at io.quarkus.deployment.steps.UndertowBuildStep$build-649634386.deploy(UndertowBuildStep$build-649634386.zig:40)
	at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:169)
	... 43 more
Caused by: java.lang.RuntimeException: java.lang.NullPointerException
	at io.undertow.servlet.core.DeploymentManagerImpl.deploy(DeploymentManagerImpl.java:253)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder.bootServletContainer(UndertowDeploymentRecorder.java:517)
	... 46 more
Caused by: java.lang.NullPointerException
	at io.quarkus.qe.undertow.PongLeakSample.contextInitialized(PongLeakSample.java:51)
```

Here I paste you a full reproducer:

```
import java.io.IOException;
import java.util.Collections;
import java.util.IdentityHashMap;
import java.util.Set;

import javax.enterprise.inject.spi.DeploymentException;
import javax.servlet.ServletContextEvent;
import javax.servlet.ServletContextListener;
import javax.servlet.annotation.WebListener;
import javax.websocket.CloseReason;
import javax.websocket.Endpoint;
import javax.websocket.EndpointConfig;
import javax.websocket.Session;
import javax.websocket.server.ServerContainer;
import javax.websocket.server.ServerEndpointConfig;

import org.jboss.logging.Logger;

@WebListener
public class PongLeakSample implements ServletContextListener {
    private static final Logger LOG = Logger.getLogger(PongLeakSample.class);

    private Thread pingThread;
    private Set<WebsockEndpoint> endpoints;

    public PongLeakSample() {
        endpoints = Collections.synchronizedSet(Collections.newSetFromMap(new IdentityHashMap<>()));
    }

    @Override
    public void contextDestroyed(ServletContextEvent sce) {
        pingThread.interrupt();
        endpoints.forEach(WebsockEndpoint::close);
        try {
            pingThread.join();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    /**
     * @see ServletContextListener#contextInitialized(ServletContextEvent)
     */
    @Override
    public void contextInitialized(ServletContextEvent sce) {
        try {
            ((ServerContainer) sce.getServletContext().getAttribute(ServerContainer.class.getName()))
                    .addEndpoint(ServerEndpointConfig.Builder.create(WebsockEndpoint.class, "/simple")
                            .configurator(new WebsockEndpointConfigurator(this)).build());
            
        } catch (DeploymentException | javax.websocket.DeploymentException e) {
            e.printStackTrace();
        }
        pingThread = new Thread(this::pingRoutine);
        pingThread.setDaemon(true);
        pingThread.start();
    }

    private void pingRoutine() {
        try {
            while (true) {
                Thread.sleep(300);
                endpoints.forEach(WebsockEndpoint::ping);
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    public static class WebsockEndpoint extends Endpoint {

        private static final String PONG = "PONG";
        private final PongLeakSample filter;
        private Session session;

        public WebsockEndpoint(PongLeakSample filter) {
            this.filter = filter;
        }

        @Override
        public void onOpen(Session session, EndpointConfig config) {
            this.session = session;
            filter.endpoints.add(this);
        }

        @Override
        public void onClose(Session session, CloseReason closeReason) {
            filter.endpoints.remove(this);
        }

        public void ping() {
            LOG.info("ping-pong invoked!");
            session.getAsyncRemote().sendText(PONG);
        }

        public void close() {
            try {
                session.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public static class WebsockEndpointConfigurator extends ServerEndpointConfig.Configurator {
        private PongLeakSample filter;

        WebsockEndpointConfigurator(PongLeakSample filter) {
            this.filter = filter;
        }

        @Override
        public <T> T getEndpointInstance(Class<T> endpointClass) throws InstantiationException {
            WebsockEndpoint endpoint = new WebsockEndpoint(filter);
            if (endpointClass.isInstance(endpoint))
                return endpointClass.cast(endpoint);
            throw new InstantiationException(
                    "Requested class is not assignable from " + WebsockEndpoint.class.getName());
        }
    }
}
```

Run: `mvn quarkus:dev`

Test: `curl --include \\n     --no-buffer \\n     --header "Connection: Upgrade" \\n     --header "Upgrade: websocket" \\n     --header "Sec-WebSocket-Key: SGVsbG8sIHdvcmxkIQ==" \\n     --header "Sec-WebSocket-Version: 13" \\n     http://localhost:8080/simple`

Works on: 1.11.7.Final

Fails on: 1.13.7.Final (and higher)

FYI you can also fetch an instance of the configured `ServerContainer` by doing:

```java
ServerContainer container = (ServerContainer) ContainerProvider.getWebSocketContainer();
```

But then you get an error when you try to deploy:

```java
Caused by: java.lang.RuntimeException: java.lang.IllegalStateException: UT003017: Cannot add endpoint after deployment
	at io.undertow.servlet.core.DeploymentManagerImpl.deploy(DeploymentManagerImpl.java:253)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder.bootServletContainer(UndertowDeploymentRecorder.java:510)
```

--- a/extensions/undertow/runtime/src/main/java/io/quarkus/undertow/runtime/UndertowDeploymentRecorder.java
+++ b/extensions/undertow/runtime/src/main/java/io/quarkus/undertow/runtime/UndertowDeploymentRecorder.java
@@ -523,7 +523,11 @@ public class UndertowDeploymentRecorder {
     }
 
     public void addServletContextAttribute(RuntimeValue<DeploymentInfo> deployment, String key, Object value1) {
-        deployment.getValue().addServletContextAttribute(key, value1);
+        if (value1 instanceof RuntimeValue) {
+            deployment.getValue().addServletContextAttribute(key, ((RuntimeValue<?>) value1).getValue());
+        } else {
+            deployment.getValue().addServletContextAttribute(key, value1);
+        }
     }
 
     public void addServletExtension(RuntimeValue<DeploymentInfo> deployment, ServletExtension extension) {

--- a/extensions/websockets/client/deployment/pom.xml
+++ b/extensions/websockets/client/deployment/pom.xml
@@ -23,6 +23,10 @@
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-undertow-spi</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
             <artifactId>quarkus-junit5-internal</artifactId>
             <scope>test</scope>
         </dependency>

--- a/extensions/websockets/client/deployment/src/main/java/io/quarkus/undertow/websockets/client/deployment/WebsocketClientProcessor.java
+++ b/extensions/websockets/client/deployment/src/main/java/io/quarkus/undertow/websockets/client/deployment/WebsocketClientProcessor.java
@@ -12,6 +12,7 @@ import javax.websocket.ClientEndpointConfig;
 import javax.websocket.ContainerProvider;
 import javax.websocket.Endpoint;
 import javax.websocket.server.ServerApplicationConfig;
+import javax.websocket.server.ServerContainer;
 import javax.websocket.server.ServerEndpointConfig;
 
 import org.jboss.jandex.AnnotationInstance;
@@ -34,10 +35,11 @@ import io.quarkus.deployment.builditem.FeatureBuildItem;
 import io.quarkus.deployment.builditem.ServiceStartBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ServiceProviderBuildItem;
-import io.quarkus.netty.deployment.EventLoopSupplierBuildItem;
 import io.quarkus.runtime.RuntimeValue;
+import io.quarkus.undertow.deployment.ServletContextAttributeBuildItem;
 import io.quarkus.undertow.websockets.client.runtime.WebsocketCoreRecorder;
 import io.undertow.websockets.DefaultContainerConfigurator;
+import io.undertow.websockets.ServerWebSocketContainer;
 import io.undertow.websockets.UndertowContainerProvider;
 import io.undertow.websockets.WebSocketDeploymentInfo;
 
@@ -70,16 +72,16 @@ public class WebsocketClientProcessor {
     }
 
     @BuildStep
-    @Record(ExecutionTime.RUNTIME_INIT)
+    @Record(ExecutionTime.STATIC_INIT)
     public ServerWebSocketContainerBuildItem deploy(final CombinedIndexBuildItem indexBuildItem,
             WebsocketCoreRecorder recorder,
             BuildProducer<ReflectiveClassBuildItem> reflection,
-            EventLoopSupplierBuildItem eventLoopSupplierBuildItem,
             List<AnnotatedWebsocketEndpointBuildItem> annotatedEndpoints,
             BeanContainerBuildItem beanContainerBuildItem,
             WebsocketConfig websocketConfig,
             BuildProducer<WebSocketDeploymentInfoBuildItem> infoBuildItemBuildProducer,
-            Optional<ServerWebSocketContainerFactoryBuildItem> factoryBuildItem) throws Exception {
+            Optional<ServerWebSocketContainerFactoryBuildItem> factoryBuildItem,
+            BuildProducer<ServletContextAttributeBuildItem> servletContextAttributeBuildItemBuildProducer) throws Exception {
 
         final Set<String> endpoints = new HashSet<>();
         final Set<String> config = new HashSet<>();
@@ -121,10 +123,14 @@ public class WebsocketClientProcessor {
                 websocketConfig.maxFrameSize,
                 websocketConfig.dispatchToWorker);
         infoBuildItemBuildProducer.produce(new WebSocketDeploymentInfoBuildItem(deploymentInfo));
+        RuntimeValue<ServerWebSocketContainer> serverContainer = recorder.createServerContainer(
+                beanContainerBuildItem.getValue(),
+                deploymentInfo,
+                factoryBuildItem.map(ServerWebSocketContainerFactoryBuildItem::getFactory).orElse(null));
+        servletContextAttributeBuildItemBuildProducer
+                .produce(new ServletContextAttributeBuildItem(ServerContainer.class.getName(), serverContainer));
         return new ServerWebSocketContainerBuildItem(
-                recorder.createServerContainer(beanContainerBuildItem.getValue(), eventLoopSupplierBuildItem.getMainSupplier(),
-                        deploymentInfo,
-                        factoryBuildItem.map(ServerWebSocketContainerFactoryBuildItem::getFactory).orElse(null)));
+                serverContainer);
     }
 
     public static void registerCodersForReflection(BuildProducer<ReflectiveClassBuildItem> reflection,

--- a/extensions/websockets/client/runtime/src/main/java/io/quarkus/undertow/websockets/client/runtime/WebsocketCoreRecorder.java
+++ b/extensions/websockets/client/runtime/src/main/java/io/quarkus/undertow/websockets/client/runtime/WebsocketCoreRecorder.java
@@ -19,6 +19,7 @@ import io.quarkus.arc.ManagedContext;
 import io.quarkus.arc.runtime.BeanContainer;
 import io.quarkus.runtime.RuntimeValue;
 import io.quarkus.runtime.annotations.Recorder;
+import io.quarkus.vertx.core.runtime.VertxCoreRecorder;
 import io.undertow.websockets.ServerWebSocketContainer;
 import io.undertow.websockets.UndertowContainerProvider;
 import io.undertow.websockets.WebSocketDeploymentInfo;
@@ -26,6 +27,7 @@ import io.undertow.websockets.util.ContextSetupHandler;
 import io.undertow.websockets.util.ObjectFactory;
 import io.undertow.websockets.util.ObjectHandle;
 import io.undertow.websockets.util.ObjectIntrospecter;
+import io.vertx.core.impl.VertxInternal;
 
 @Recorder
 public class WebsocketCoreRecorder {
@@ -105,7 +107,6 @@ public class WebsocketCoreRecorder {
     }
 
     public RuntimeValue<ServerWebSocketContainer> createServerContainer(BeanContainer beanContainer,
-            Supplier<EventLoopGroup> eventLoopGroupSupplier,
             RuntimeValue<WebSocketDeploymentInfo> infoVal, ServerWebSocketContainerFactory serverContainerFactory)
             throws DeploymentException {
         WebSocketDeploymentInfo info = infoVal.getValue();
@@ -136,7 +137,12 @@ public class WebsocketCoreRecorder {
                     }
                 };
             }
-        }, Thread.currentThread().getContextClassLoader(), eventLoopGroupSupplier,
+        }, Thread.currentThread().getContextClassLoader(), new Supplier<EventLoopGroup>() {
+            @Override
+            public EventLoopGroup get() {
+                return ((VertxInternal) VertxCoreRecorder.getVertx().get()).getEventLoopGroup();
+            }
+        },
                 Collections.singletonList(new ContextSetupHandler() {
                     @Override
                     public <T, C> Action<T, C> create(Action<T, C> action) {

--- a/integration-tests/main/src/main/java/io/quarkus/it/websocket/AddWebSocketHandler.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/websocket/AddWebSocketHandler.java
@@ -1,0 +1,41 @@
+package io.quarkus.it.websocket;
+
+import javax.enterprise.inject.spi.DeploymentException;
+import javax.servlet.ServletContextEvent;
+import javax.servlet.ServletContextListener;
+import javax.servlet.annotation.WebListener;
+import javax.websocket.Endpoint;
+import javax.websocket.EndpointConfig;
+import javax.websocket.Session;
+import javax.websocket.server.ServerContainer;
+import javax.websocket.server.ServerEndpointConfig;
+
+import io.quarkus.runtime.annotations.RegisterForReflection;
+
+@WebListener
+public class AddWebSocketHandler implements ServletContextListener {
+    @Override
+    public void contextDestroyed(ServletContextEvent sce) {
+    }
+
+    @Override
+    public void contextInitialized(ServletContextEvent sce) {
+        try {
+            ((ServerContainer) sce.getServletContext().getAttribute(ServerContainer.class.getName()))
+                    .addEndpoint(ServerEndpointConfig.Builder.create(WebSocketEndpoint.class, "/added-dynamic").build());
+
+        } catch (DeploymentException | javax.websocket.DeploymentException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @RegisterForReflection
+    public static class WebSocketEndpoint extends Endpoint {
+
+        @Override
+        public void onOpen(Session session, EndpointConfig config) {
+            session.getAsyncRemote().sendText("DYNAMIC");
+        }
+
+    }
+}

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/WebsocketTestCase.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/WebsocketTestCase.java
@@ -32,6 +32,8 @@ public class WebsocketTestCase {
 
     @TestHTTPResource("wsopen")
     URI openURI;
+    @TestHTTPResource("added-dynamic")
+    URI added;
 
     @Test
     public void websocketTest() throws Exception {
@@ -52,6 +54,29 @@ public class WebsocketTestCase {
 
         try {
             Assertions.assertEquals("hello", message.poll(20, TimeUnit.SECONDS));
+        } finally {
+            session.close();
+        }
+    }
+
+    @Test
+    public void addedWebSocketTest() throws Exception {
+
+        LinkedBlockingDeque<String> message = new LinkedBlockingDeque<>();
+        Session session = ContainerProvider.getWebSocketContainer().connectToServer(new Endpoint() {
+            @Override
+            public void onOpen(Session session, EndpointConfig endpointConfig) {
+                session.addMessageHandler(new MessageHandler.Whole<String>() {
+                    @Override
+                    public void onMessage(String s) {
+                        message.add(s);
+                    }
+                });
+            }
+        }, ClientEndpointConfig.Builder.create().build(), added);
+
+        try {
+            Assertions.assertEquals("DYNAMIC", message.poll(20, TimeUnit.SECONDS));
         } finally {
             session.close();
         }

